src/components/Pacs/PacsController.tsx

```diff
@@ -0,0 +1,660 @@
+ * 5. ChRIS_ui checks CUBE whether a series exists in CUBE
+ * 6. When both subscription and existence check is complete,
+ *    and the series does not exist in CUBE, ChRIS_ui is ready
+ *    to pull the DICOM series.
+ * 7. During the reception of a DICOM series, `status.done === false`
+ * 8. After the reception of a DICOM series, ChRIS enters a "waiting"
+ *    state while the task to register the DICOM series is enqueued
+ *    or running.
+ * 9. The DICOM series will appear in CUBE after being registered.
+ */
+const PacsController: React.FC<PacsControllerProps> = ({
+  getChrisClient,
+  getPfdcmClient,
+}) => {
+  // ========================================
+  // CLIENTS AND MISC
+  // ========================================
+
+  const { message } = App.useApp();
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: https://github.com/biomejs/biome/issues/4248
+  const pfdcmClient = React.useMemo(getPfdcmClient, [getPfdcmClient]);
+  // biome-ignore lint/correctness/useExhaustiveDependencies: https://github.com/biomejs/biome/issues/4248
+  const chrisClient = React.useMemo(getChrisClient, [getChrisClient]);
+
+  // ========================================
+  // STATE
+  // ========================================
+
+  /**
+   * Indicates a fatal error with the WebSocket.
+   */
+  const [wsError, setWsError] = React.useState<React.ReactNode | null>(null);
+  // TODO create a settings component for changing preferences
+  const [preferences, setPreferences] = React.useState(DEFAULT_PREFERENCES);
+
+  /**
+   * The state of DICOM series, according to LONK.
+   */
+  const [receiveState, setReceiveState] = useImmer<ReceiveState>(
+    new SeriesMap(),
+  );
+
+  /**
+   * List of PACS queries which the user wants to pull.
+   */
+  const [pullRequests, setPullRequests] = useImmer<PullRequestStates>(
+    new Map(),
+  );
+
+  /**
+   * Update the state of a pull request.
+   */
+  const updatePullRequestState = React.useCallback(
+    (query: SpecificDicomQuery, delta: Partial<PacsPullRequestState>) =>
+      setPullRequests((draft) => {
+        const prev = draft.get(query);
+        if (!prev) {
+          throw new Error(
+            "pullFromPacs mutation called on unknown pull request: " +
+              `service=${service}, query=${JSON.stringify(query)}`,
+          );
+        }
+        draft.set(query, { ...prev, ...delta });
+      }),
+    [setPullRequests],
+  );
+
+  // ========================================
+  // PFDCM QUERIES AND DATA
+  // ========================================
+
+  const [{ service, query }, setPacsQuery] = React.useState<{
+    service?: string;
+    query?: PACSqueryCore;
+  }>({});
+
+  /**
+   * List of PACS servers which PFDCM can talk to.
+   */
+  const pfdcmServices = useQuery({
+    queryKey: ["pfdcmServices"],
+    queryFn: () => pfdcmClient.getPacsServices(),
+  });
+
+  /**
+   * The state of DICOM studies and series in PFDCM.
+   */
+  const pfdcmStudies = useQuery({
```

Comment:
React Query internally manages the comparison of queryKey contents, wrapping it with useMemo to prevent a new array instance could be redudant?